subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc

```diff
@@ -141,6 +141,54 @@ These errors can be resolved by updating the violating code or dependency. Updat
 
 For some examples of possible error or warning outputs which may arise due to this change, see <<remove_test_add_opens>>.
 
+==== SourceSet classesDirs no longer depends upon the entire SourceSet as a task dependency
+
+Prior to Gradle 8.0, the task dependencies for the `classesDirs` of a given `SourceSetOutput`
+included the `SourceSetOutput` itself as a dependency. This meant that directly
+computing `classesDirs` would also execute `classes`, `processResources`, and any other
+task dependency for the `SourceSetOutput`.
+Since 8.0, this implicit dependency is removed. Now, computing `classesDirs` only executes the
+tasks which directly contribute to it and none of the tasks which contribute to the parent `SourceSetOutput`.
+
+Consider the following buildscript:
+```groovy
+plugins {
+    id 'java-library'
+}
+def fooTask = tasks.register("foo") {
+    outputs.file(project.layout.buildDirectory.dir("fooOut"))
+}
+tasks.register("bar") {
+    inputs.files(java.sourceSets.main.output.classesDirs)
+}
+java {
+    sourceSets {
+        main {
+            resources.srcDir(fooTask.map { it.outputs.files.singleFile })
+        }
+    }
+}
```

Comment:
I think I would focus this on just the new behavior of classesDirs and with a concrete kind of example.  It's a little disconnected that the fooTask isn't executed. It's not executed because processResources isn't executed.

```suggestion
// Task lists all files in the given classFiles FileCollection
tasks.register("listClassFiles", ListClassFiles) {
    classFiles.from(java.sourceSets.main.output.classesDirs)
}
```